### PR TITLE
many: add `refresh-inhibit` notice

### PR DESCRIPTION
**Pull request by ZeyadYasser**
This PR adds `refresh-inhibit` notice which is needed for RAA UX flow as mentioned in [RAA-UX spec SD167](https://docs.google.com/document/d/1HJQWKzgaB3yPKwRUqlxYhgyaKG5IcJe8eZOT4YY1CI8/edit#heading=h.c40tauj2e3fo).

Once https://github.com/snapcore/snapd/pull/13576 is merged, more complex state-based decisions can be made, like only sending when the list of inhibited snaps changes.

This commit https://github.com/snapcore/snapd/commit/1e6dec8ec6e77e84be453d85c9e05dbc308965d7 can be split into a different PR if needed.

### CODE REVIEW

File: overlord/snapstate/autorefresh.go [hunk 1]
```diff
@@ -797,6 +810,37 @@ func inhibitRefresh(st *state.State, snapst *SnapState, snapsup *SnapSetup, info
...
+	if err != nil {
+		return err
+	}
+
+	for _, snapst := range snapStates {
+		if snapst.RefreshInhibitedTime == nil {
+			continue
+		}
+		// One inhibited snap is enough to send the notice.
+		if _, err := st.AddNotice(nil, state.RefreshInhibitNotice, "-", nil); err != nil {
```
Comment (olivercalder) [inline overlord/snapstate/autorefresh.go:835]: Is there any notice issued when the snap becomes uninhibited? If the key is always `"-"`, it seems like there would be no way to know. If there were a key for "inhibited" and a key for "uninhibited", then whichever one was more recent would be the current inhibition state.

Alternatively, this could be encoded in the notice data.

But perhaps we don't want to have this kind of information. If that is the case, why not?

Comment (olivercalder) [reply overlord/snapstate/autorefresh.go:835]: Ahh, probably because of race conditions. Need to check inhibited status and get lock directly, in case it becomes {,un}inhibited between when the notice is received and when that information is used. So the notice then is just saying "inhibited status changed, go query it". Am I understanding correctly?

**Review by olivercalder — COMMENTED** [review]
Looks good, just a few questions. Thanks!

File: overlord/snapstate/autorefresh.go (hunk 1 above)
Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh.go:835]: In the spec we don't need to track snaps getting uninhibited, only in very special cases when a refresh is forced, and in that case a change-update notice + api query can be used as mentioned in the spec.

The intention behind this notice is exactly like you mentioned "inhibited status changed, go query it" specifically querying [refresh-inhibit-proceed-time](https://github.com/snapcore/snapd/pull/13580) `/v2/snaps` . "-" is used intentionally (because it is not a valid snap-name) to indicate that this notice is not about a specific snap, allowing us to later use this key to indicate inhibition status  changes per-snap if we ever need to do that.

File: overlord/state/notices_test.go [hunk 2]
```diff
@@ -297,12 +299,21 @@ func (s *noticesSuite) TestNoticesFilterType(c *C) {
...
+	c.Check(n["key"], Equals, "-")
+
 	// Multiple types
 	notices = st.Notices(&state.NoticeFilter{Types: []state.NoticeType{
 		state.ChangeUpdateNotice,
 		state.WarningNotice,
+		state.RefreshInhibitNotice,
 	}})
-	c.Assert(notices, HasLen, 4)
+	c.Assert(notices, HasLen, 5)
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply overlord/state/notices_test.go]: Good point, thanks! updated.

File: overlord/snapstate/snapmgr.go [hunk 3]
```diff
@@ -682,9 +684,21 @@ func (m *SnapManager) StartUp() error {
 		return fmt.Errorf("failed to generate cookies: %q", err)
 	}
 
+	m.changeCallbackID = m.state.AddChangeStatusChangedHandler(processInhibitedAutoRefresh)
```
Comment (pedronis) [inline overlord/snapstate/snapmgr.go:687]: this merits a comment about what is being done here? also can you explain why you chose to set up things this way and have explicit emittting in a few places?

File: overlord/snapstate/autorefresh.go [hunk 4]
```diff
@@ -797,6 +810,37 @@ func inhibitRefresh(st *state.State, snapst *SnapState, snapsup *SnapSetup, info
...
+	}
+}
+
+func maybeAddRefreshInhibitNotice(st *state.State) error {
+	snapStates, err := All(st)
+	if err != nil {
+		return err
+	}
+
+	for _, snapst := range snapStates {
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go:830]: this ultimately would need to remember the set of snaps, and emit only if the set has changed? has we don't have RepeatCheck anymore we need anyway to roll something here explicitly, the PR is small enough that we should do that now

**Review by pedronis — COMMENTED** [review]
thanks, some comments/questions

File: overlord/snapstate/snapmgr.go (hunk 3 above)
Comment (ZeyadYasser) [reply overlord/snapstate/snapmgr.go:687]: The only other entrypoint I can think of that can guarantee that all snaps were checked is the `check-rerefresh` handler, it would be something like this:
```diff
diff --git a/overlord/snapstate/handlers.go b/overlord/snapstate/handlers.go
index c49079088b..5e391f43ad 100644
--- a/overlord/snapstate/handlers.go
+++ b/overlord/snapstate/handlers.go
@@ -4356,6 +4356,11 @@ func (m *SnapManager) doCheckReRefresh(t *state.Task, tomb *tomb.Tomb) error {
                return &state.Retry{After: reRefreshRetryTimeout, Reason: "pending refreshes"}
        }
 
+       chg := t.Change()
+       if chg.Kind() == "auto-refresh" {
+               defer maybeAddRefreshInhibitNotice(st)
+       }
+
        snaps, failed, err := refreshedSnaps(t)
        if err != nil {
                return err
@@ -4366,8 +4371,6 @@ func (m *SnapManager) doCheckReRefresh(t *state.Task, tomb *tomb.Tomb) error {
                }
        }
 
-       chg := t.Change()
-
        // if any snap failed to refresh, reconsider validation set tracking
        if failed {
                tasksets, err := maybeRestoreValidationSetsAndRevertSnaps(st, snaps, chg.ID())
```

but we would still need the pre-dl check in `launchAutoRefresh` for the case that all snaps are inhibited early in the soft check.

The main requirement for the emitting is being able to observe the overall auto-refresh change status for all snaps. The change status change handler seemed like a natural place for this.

I don't have strong preferences here, would love to know if you have other entrypoints in mind?

File: overlord/snapstate/autorefresh.go [hunk 5]
```diff
@@ -797,6 +810,37 @@ func inhibitRefresh(st *state.State, snapst *SnapState, snapsup *SnapSetup, info
 	return busyErr
 }
 
+func processInhibitedAutoRefresh(chg *state.Change, old state.Status, new state.Status) {
+	if !new.Ready() || chg.Kind() != "auto-refresh" {
```
[the remark this answers is not in the document]
Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh.go]: Good point, Thank you! updated

File: overlord/snapstate/snapmgr.go (hunk 3 above)
Comment (ZeyadYasser) [reply overlord/snapstate/snapmgr.go:689]: Added a comment explaining what was being registered and why, thanks!

File: overlord/snapstate/autorefresh.go (hunk 4 above)
Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh.go:836]: updated, thanks!

Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh.go:836]: I went with a simple mechanism of saving the last set of inhibited snaps in state in `last-recorded-inhibited-snaps`. I don't have a strong preference on the key name.

File: overlord/snapstate/snapstate_update_test.go [hunk 6]
```diff
@@ -10287,6 +10236,334 @@ func (s *snapmgrTestSuite) TestAutoRefreshCreatePreDownload(c *C) {
...
+	snapstate.Set(s.state, "some-other-snap", &snapstate.SnapState{
+		Active: true,
+		Sequence: snapstatetest.NewSequenceFromSnapSideInfos([]*snap.SideInfo{
+			{RealName: "some-other-snap", SnapID: "some-other-snap-id", Revision: snap.R(1)},
+		}),
+		Current:  snap.R(1),
+		SnapType: string(snap.TypeApp),
+	})
+
+	// Force another auto-refresh to check that notice is recorded becuase
```
Comment (olivercalder) [inline overlord/snapstate/snapstate_update_test.go:10515]: As the misspell checker mentions, `s/becuase/because/` :)

File: overlord/state/notices.go [hunk 7]
```diff
@@ -190,11 +190,14 @@ const (
...
 	// warning message.
 	WarningNotice NoticeType = "warning"
+
+	// Recorded whenever an auto-refresh is inhibited for one or more snaps.
+	RefreshInhibitNotice NoticeType = "refresh-inhibit"
 )
 
 func (t NoticeType) Valid() bool {
 	switch t {
-	case ChangeUpdateNotice, WarningNotice:
```
Comment (olivercalder) [inline overlord/state/notices.go:197]: I believe this will conflict with this from #13619, depending on which is merged first.

https://github.com/snapcore/snapd/blob/ebca9777aadd66d72c46bd17a6985d82e566cc9c/overlord/state/notices.go#L201-L209

**Review by olivercalder — APPROVED** [review]
Looks good to me, thank you!

File: overlord/snapstate/autorefresh_gating_test.go [hunk 8]
```diff
@@ -1950,7 +1950,6 @@ func (s *snapmgrTestSuite) testAutoRefreshPhase2(c *C, beforePhase1 func(), gate
 		chg.AddAll(ts)
 	}
 
-	defer s.se.Stop()
```
Comment (pedronis) [inline overlord/snapstate/autorefresh_gating_test.go:1953]: why these changes?

File: overlord/snapstate/autorefresh.go [hunk 9]
```diff
@@ -797,6 +810,58 @@ func inhibitRefresh(st *state.State, snapst *SnapState, snapsup *SnapSetup, info
...
+			if !lastRecordedInhibitedSnaps[snapName] {
+				changed = true
+				break
+			}
+		}
+	}
+
+	if changed {
+		// XXX: do we want to record a notice if no snaps are inhibited (i.e. the set
+		// of inhibited snaps changed to an empty set)?
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go:855]: I think so

File: overlord/snapstate/autorefresh.go [hunk 10]
```diff
@@ -644,11 +644,22 @@ func (m *autoRefresh) launchAutoRefresh() error {
...
 	if len(updateTss.Refresh) == 0 {
+		if createdPreDl {
+			// NOTE: If all refresh candidates are blocked from auto-refresh by checks
+			// in softCheckNothingRunningForRefresh then no auto-refresh change will be
+			// created (i.e. len(updateTss.Refresh) == 0) and only a pre-download change
+			// is created for those snaps. This still means that some snaps are being
+			// inhibited from auto-refresh so we are recording a notice about it here
+			// because it cannot be captured in processInhibitedAutoRefresh which only
+			// looks for auto-refresh changes.
+			return maybeAddRefreshInhibitNotice(m.state)
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go:661]: which means this maybe needs to uncoditional now? and always be called if createPrDl is true or false

**Review by pedronis — COMMENTED** [review]
some comments/questions

File: overlord/snapstate/autorefresh_gating_test.go (hunk 8 above)
Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh_gating_test.go:1953]: As discussed with @Meulengracht `Stop()` in managers must be called with state unlocked, since `SnapManager` didn't implement `Stop()` before this PR, it was not caught earlier and the tests were able to get away with doing that. To fix this I am calling `Stop()` on test cleanup https://github.com/snapcore/snapd/commit/1e6dec8ec6e77e84be453d85c9e05dbc308965d7, this commit can be extracted into a separate PR, but since it was blocking my tests I added it here.

File: overlord/snapstate/snapstate_update_test.go (hunk 6 above)
Comment (ZeyadYasser) [reply overlord/snapstate/snapstate_update_test.go]: Thank you! updated

File: overlord/state/notices.go (hunk 7 above)
Comment (ZeyadYasser) [reply overlord/state/notices.go:197]: Thanks for reminding! I think I also need to account for the type checker for `refresh-inhibit`, depending on which is merged first.

File: overlord/snapstate/autorefresh.go (hunk 9 above)
Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh.go]: comment removed

File: overlord/snapstate/autorefresh.go (hunk 10 above)
Comment (ZeyadYasser) [reply overlord/snapstate/autorefresh.go]: Now that `maybeAddRefreshInhibitNotice` checks last recorded inhibited snaps set, the condition is a no-op. I removed it, Thank you!

**Review by pedronis — APPROVED** [review]
thank you

**Review by Meulengracht — APPROVED** [review]
Thanks, just a comment
